File: handler/group.go
package handler

import (
	"github.com/MXWXZ/skynet/sn"
	"github.com/google/uuid"
	"gorm.io/gorm"
)

type GroupImpl struct {
	tx    *gorm.DB
	group *sn.ORM[sn.Group]
	link  *sn.ORM[sn.UserGroupLink]
}

func NewGroupHandler() sn.GroupHandler {
	return &GroupImpl{
		tx:    sn.Skynet.DB,
		group: sn.NewORM[sn.Group](sn.Skynet.DB),
		link:  sn.NewORM[sn.UserGroupLink](sn.Skynet.DB),
	}
}

func (u *GroupImpl) WithTx(tx *gorm.DB) sn.GroupHandler {
	return &GroupImpl{
		tx:    tx,
		group: sn.NewORM[sn.Group](tx),
		link:  sn.NewORM[sn.UserGroupLink](tx),
	}
}

func (u *GroupImpl) New(name string, note string) (*sn.Group, error) {
	group := &sn.Group{
		Name: name,
		Note: note,
	}
	if err := u.group.Create(group); err != nil {
		return nil, err
	}
	return group, nil
}

func (u *GroupImpl) Link(uid []uuid.UUID, gid []uuid.UUID) ([]*sn.UserGroupLink, error) {
	if len(uid) == 0 || len(gid) == 0 {
		return nil, nil
	}
	var group []*sn.UserGroupLink
	for _, v := range uid {
		for _, g := range gid {
			group = append(group, &sn.UserGroupLink{
				UID: v,
				GID: g,
			})
		}
	}
	if err := u.link.Creates(group); err != nil {
		return nil, err
	}
	return group, nil
}

func (impl *GroupImpl) GetGroupAllUser(id uuid.UUID, cond *sn.Condition) ([]*sn.UserGroupLink, error) {
	link, err := impl.link.Where("user_group_links.gid = ?", id).Cond(cond).Joins("User").Find()
	if err != nil {
		return nil, err
	}
	return link, nil
}

func (impl *GroupImpl) CountGroupAllUser(id uuid.UUID, cond *sn.Condition) (int64, error) {
	return impl.link.Where("user_group_links.gid = ?", id).Count(cond)
}

func (impl *GroupImpl) GetAll(cond *sn.Condition) ([]*sn.Group, error) {
	return impl.group.Cond(cond).Find()
}

func (impl *GroupImpl) Get(id uuid.UUID) (*sn.Group, error) {
	return impl.group.Take(id)
}

func (impl *GroupImpl) GetByName(name string) (*sn.Group, error) {
	return impl.group.Where("name = ?", name).Take()
}

func (impl *GroupImpl) GetUserAllGroup(id uuid.UUID) ([]*sn.UserGroupLink, error) {
	link, err := impl.link.Where("user_group_links.uid = ?", id).Joins("Group").Find()
	if err != nil {
		return nil, err
	}
	return link, nil
}

func (impl *GroupImpl) Count(cond *sn.Condition) (int64, error) {
	return impl.group.Count(cond)
}

func (impl *GroupImpl) Update(column []string, group *sn.Group) error {
	if group == nil {
		return nil
	}
	return impl.group.ID(group.ID).Updates(column, group)
}

func (impl *GroupImpl) Unlink(uid []uuid.UUID, gid []uuid.UUID) error {
	if len(uid) == 0 && len(gid) == 0 {
		return nil
	}
	return impl.link.Transaction(func(tx *gorm.DB) error {
		if uid != nil && gid != nil {
			for _, i := range uid {
				for _, j := range gid {
					if _, err := impl.link.WithTx(tx).Delete("uid = ? AND gid = ?", i, j); err != nil {
						return err
					}
				}
			}
		} else if uid == nil {
			for _, j := range gid {
				if _, err := impl.link.WithTx(tx).Delete("gid = ?", j); err != nil {
					return err
				}
			}
		} else {
			for _, i := range uid {
				if _, err := impl.link.WithTx(tx).Delete("uid = ?", i); err != nil {
					return err
				}
			}
		}
		return nil
	})
}

func (impl *GroupImpl) Delete(id uuid.UUID) error {
	if _, err := impl.group.DeleteID(id); err != nil {
		return err
	}
	return nil
}

File: skynet/src/handler_impl/group.rs
use anyhow::Result;
use async_trait::async_trait;
use derivative::Derivative;
use sea_orm::{
    ActiveModelBehavior, ActiveModelTrait, ActiveValue::NotSet, ColumnTrait, DatabaseTransaction,
    EntityTrait, JoinType::InnerJoin, ModelTrait, PaginatorTrait, QueryFilter, QuerySelect,
    RelationTrait, Set, Unchanged,
};
use skynet::{
    entity::{groups, user_group_links, users},
    handler::GroupHandler,
    hyuuid::uuid2string,
    Condition, HyUuid,
};
use skynet_macro::default_handler_impl;

#[derive(Derivative)]
#[derivative(Default(new = "true"), Debug)]
pub struct DefaultGroupHandler;

#[default_handler_impl(groups)]
#[async_trait]
impl GroupHandler for DefaultGroupHandler {
    async fn find_user_group(
        &self,
        db: &DatabaseTransaction,
        uid: &HyUuid,
        join: bool,
    ) -> Result<Vec<groups::Model>> {
        if join {
            users::Model {
                id: uid.to_owned(),
                ..Default::default()
            }
            .find_linked(user_group_links::UserToGroup)
            .all(db)
            .await
            .map_err(anyhow::Error::from)
        } else {
            let mut ret = Vec::new();
            let group = user_group_links::Entity::find()
                .filter(user_group_links::Column::Uid.eq(*uid))
                .all(db)
                .await?;
            for i in group {
                ret.push(groups::Model {
                    id: i.gid,
                    ..Default::default()
                });
            }
            Ok(ret)
        }
    }

    async fn find_group_user(
        &self,
        db: &DatabaseTransaction,
        gid: &HyUuid,
        cond: Condition,
    ) -> Result<(Vec<users::Model>, u64)> {
        let (q, page) = cond
            .add(user_group_links::Column::Gid.eq(*gid))
            .build(users::Entity::find());
        let q = q.join(InnerJoin, user_group_links::Relation::User.def().rev());
        if let Some(page) = page {
            let q = q.paginate(db, page.size);
            Ok((q.fetch_page(page.page - 1).await?, q.num_items().await?))
        } else {
            let res = q.all(db).await?;
            let cnt = res.len() as u64;
            Ok((res, cnt))
        }
    }

    async fn find_group_user_by_id(
        &self,
        db: &DatabaseTransaction,
        gid: &HyUuid,
        uid: &HyUuid,
    ) -> Result<Option<users::Model>> {
        users::Entity::find()
            .filter(user_group_links::Column::Uid.eq(*uid))
            .filter(user_group_links::Column::Gid.eq(*gid))
            .join(InnerJoin, user_group_links::Relation::User.def().rev())
            .one(db)
            .await
            .map_err(anyhow::Error::from)
    }

    async fn find_by_name(
        &self,
        db: &DatabaseTransaction,
        name: &str,
    ) -> Result<Option<groups::Model>> {
        groups::Entity::find()
            .filter(groups::Column::Name.eq(name))
            .one(db)
            .await
            .map_err(anyhow::Error::from)
    }

    async fn update(
        &self,
        db: &DatabaseTransaction,
        gid: &HyUuid,
        name: Option<&str>,
        note: Option<&str>,
    ) -> Result<groups::Model> {
        groups::ActiveModel {
            id: Unchanged(gid.to_owned()),
            name: name.map_or(NotSet, |x| Set(x.to_owned())),
            note: note.map_or(NotSet, |x| Set(x.to_owned())),
            ..Default::default()
        }
        .update(db)
        .await
        .map_err(anyhow::Error::from)
    }

    async fn create(
        &self,
        db: &DatabaseTransaction,
        name: &str,
        note: &str,
    ) -> Result<groups::Model> {
        groups::ActiveModel {
            name: Set(name.to_owned()),
            note: Set(note.to_owned()),
            ..Default::default()
        }
        .insert(db)
        .await
        .map_err(anyhow::Error::from)
    }

    async fn link(&self, db: &DatabaseTransaction, uid: &[HyUuid], gid: &[HyUuid]) -> Result<()> {
        if uid.is_empty() || gid.is_empty() {
            return Ok(());
        }
        let mut ins = Vec::new();
        for i in uid {
            for j in gid {
                ins.push(
                    user_group_links::ActiveModel {
                        uid: Set(i.to_owned()),
                        gid: Set(j.to_owned()),
                        ..Default::default()
                    }
                    .before_save(db, true) // not invoke for batch insert
                    .await?,
                );
            }
        }
        user_group_links::Entity::insert_many(ins).exec(db).await?;
        Ok(())
    }

    async fn unlink(
        &self,
        db: &DatabaseTransaction,
        uid: &[HyUuid],
        gid: &[HyUuid],
    ) -> Result<u64> {
        if uid.is_empty() && gid.is_empty() {
            Ok(0)
        } else {
            let res = if uid.is_empty() {
                user_group_links::Entity::delete_many()
                    .filter(user_group_links::Column::Gid.is_in(uuid2string(gid)))
                    .exec(db)
                    .await?
            } else if gid.is_empty() {
                user_group_links::Entity::delete_many()
                    .filter(user_group_links::Column::Uid.is_in(uuid2string(uid)))
                    .exec(db)
                    .await?
            } else {
                user_group_links::Entity::delete_many()
                    .filter(user_group_links::Column::Uid.is_in(uuid2string(uid)))
                    .filter(user_group_links::Column::Gid.is_in(uuid2string(gid)))
                    .exec(db)
                    .await?
            };
            Ok(res.rows_affected)
        }
    }
}
